db/revision_cache_interface.go

```diff
@@ -137,21 +132,64 @@ func (rev *DocumentRevision) MutableBody() (b Body, err error) {
 // If an unmarshalled copy is not available in the document revision, it makes a copy from the raw body
 // bytes and stores it in document revision itself before returning the body.
 func (rev *DocumentRevision) Body() (b Body, err error) {
-	// if we already have an unmarshalled body, take a copy and return it
-	if rev._shallowCopyBody != nil {
-		return rev._shallowCopyBody, nil
-	}
 
 	if err := b.Unmarshal(rev.BodyBytes); err != nil {
 		return nil, err
 	}
 
-	// store a copy of the unmarshalled body for next time we need it
-	rev._shallowCopyBody = b
-
 	return b, nil
 }
 
+// Inject1xBodyProperties will inject special properties (_rev etc) into document body avoiding unnecessary marshal work
```

Comment:
Do you think that there is sufficient test coverage for this? This looks like new logic to me.

I expect it would be mostly covered by 1x rest tests, but I am not always sure we actually test for all properties in each test due to need to change `_rev`.

Reply:
`TestChangesIncludeDocs` seems to test it pretty well, but will get a test that directly calls the functions and ensuring all code paths in function work as expected. The logic was copied form what https://github.com/couchbase/sync_gateway/blob/59ff006805437ce8e6bf727704c47263aaac2784/db/revision_cache_interface.go#L210 does. Implementing this way avoids the extra unmarshal and marshal work. 

Added `TestInject1xBodyProperties` for this.